File: 2022/src/nim/dayOne.nim
# Copyright (C) 2023 Rémy Cases
# See LICENSE file for extended copyright information.
# This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

import os, parseutils, std/algorithm

const bufferLen: int = 3
var bufferInt: array[bufferLen, int] = [0, 0, 0]

proc addNewInt(newValue: int) =
    if newValue > bufferInt[0]:
        bufferInt[0] = newValue
        bufferInt.sort()

proc parse(line: string, calorieCount: var int) =
    if line.len == 0: return
    var i = 0
    i.inc parseInt(line, caloriecount, i)

proc readFileAndComputeCalories(filename: string, option = '1') =
    var calorieCount = 0
    var calorieCumul = 0
    for line in filename.lines:
        if line.len == 0:
            addNewInt(calorieCumul)
            calorieCumul = 0
        else:
            parse(line, calorieCount)
            calorieCumul += calorieCount

    var calorieTotal = 0
    case option:
    of '1':
        calorieTotal = bufferInt[bufferLen - 1]
        echo "MAX CALORIES: ", calorieTotal
    of '2':
        for i in bufferInt:
            calorieTotal += i
        echo "MAX CALORIES: ", calorieTotal
    else:
        return

proc main*(part: char) =
    const file = "2022/data/input_day_one"
    let filename = getCurrentDir() / file
    readFileAndComputeCalories(filename, part)

when isMainModule:
    main('1')

File: 2022/src/rust/day_one.rs
// Copyright (C) 2023 Rémy Cases
// See LICENSE file for extended copyright information.
// This file is part of adventOfCode project from https://github.com/remyCases/adventOfCode.

use std::path::Path;
use std::env;
use std::io::{Error, ErrorKind};
use nom::*;
use nom::error::{Error as NomError};
use aoc_utils::*;

const BUFFER_LEN: usize = 3;

fn read_file_and_compute_calories(file_path: &Path, part: argparse::ArgPart) -> io::Result<()> {

    let mut calories_buf: [i32; BUFFER_LEN] = [0; BUFFER_LEN];
    let mut sum = 0;

    parse_compute!(
        file_path,
        branch::alt((
            combinator::map(
                combinator::map_res(
                    character::complete::digit0::<_, NomError<_>>,
                    str::parse
                ),
                Some
            ),
            combinator::map(
                character::complete::multispace0, 
                |_x| None
            )
        ))
        ,
        |o: Option<i32>| if let Some(x) = o {
            sum += x;
            if calories_buf[0] < sum {
                calories_buf[0] = sum;
                calories_buf.sort();
            }
        } else {
            sum = 0;
        }
    );
        
    println!("MAX CALORIES: {:}", match part {
        argparse::ArgPart::PartOne => calories_buf[calories_buf.len() - 1],
        argparse::ArgPart::PartTwo => calories_buf.iter().sum()
    });
    Ok(())
}

pub fn main(part: argparse::ArgPart) -> io::Result<()> {
    let filename = env::current_dir()?.join("2022").join("data").join("input_day_one");
    read_file_and_compute_calories(&filename, part)?;
    Ok(())
}
